index-template.md

```diff
@@ -0,0 +1,80 @@
+# OCI Index Template Protocol
+
+This is version 1 of this specification.
+
+The index-template protocol is configured via a single [URI Template][rfc6570].
+When configured via a [`refEngines` entry](ref-engine-discovery.md#ref-engines-objects), the `uri` property MUST be set, and its value is the URI Template.
+
+Consumers MUST provide at least the following variables:
+
+* `name`, matching `dns-compatible-image-name` in the [`dns-compatible-image-name` rule](dns-compatible-image-names.md).
+* `host`, matching `host` in the `dns-compatible-image-name` rule.
+* `path`, matching `path-rootless` in the `dns-compatible-image-name` rule.
+* `fragment`, matching `fragment` in the `dns-compatible-image-name` rule.
+    If `fragment` was not provided in the image name, it defaults to an empty string.
+
+and expand the URI Template as defined in [RFC 6570 section 3][rfc6570-s3].
+
+The server providing the expanded URI MUST support requests for media type [`application/vnd.oci.image.index.v1+json`][index].
+Servers MAY support other media types using HTTP content negotiation, as described in [RFC 7231 section 3.4][rfc7231-s3.4] (which is [also supported over HTTP/2][rfc7540-s8]).
+
+Consumers retrieving `application/vnd.oci.image.index.v1+json` SHOULD process it like a [layout's `index.json`][index.json], respecting [`org.opencontainers.image.ref.name` and other annotations which are recommended for `index.json`][annotations].
+
+## Example
+
+An example [`refEngines` entry](ref-engine-discovery.md#ref-engines-objects) using the [registered `oci-index-template-v1` protocol identifier](ref-engine-protocols.md) is:
+
+```json
+{
+  "protocol": "oci-index-template-v1",
+  "uri": "https://{host}/ref/{host}/{path-rootless}"
+}
+```
+
+An image name like `a.b.example.com/c/d#1.0` matches [`dns-compatible-image-name`](dns-compatible-image-names.md) with `a.b.example.com` as `host`, `c/d` as `path-rootless`, and `1.0` as `fragment` so the expanded URI is:
+
+    https://a.b.example.com/ref/a.b.example.com/c/d
+
+Retrieving that URI (with a pretend result, since [`example.com` is reserved][rfc2606-s3]):
+
+```
+$ curl -H 'Accept: application/vnd.oci.image.index.v1+json' https://a.b.example.com/ref/a.b.example.com/c/d
+{
```

Comment:
This is pure approach how to retrieve blob from image HUB. This document need only protocol, and present a object (descriptor + CAS Engine). If this need to be involved in document? IMO.

Reply:
> This is pure approach how to retrieve blob from image HUB.

I thought a full example would be informative.  The normative portion of the index template protocol spec all happens above the “Example” header.

> This document need only protocol, and present a object (descriptor + CAS Engine).

Do you mean “why is `casEngines` missing from the response?”?  I've added it with 292435d → 2bb26ec.

Reply:
>> This is pure approach how to retrieve blob from image HUB... If this need to be involved in document?

> I thought a full example would be informative.  The normative portion of the index template protocol spec all happens above the “Example” header.

I mean:
1. the object retrieved is not `application/vnd.oci.image.index.v1+json` type, it should be individual type created by discovery implementation.
2. Example just paste the object content, needn't past `curl` command. The approaches are too widely.

Reply:
>> This document need only protocol, and present a object (descriptor + CAS Engine).

> Do you mean “why is `casEngines` missing from the response?”?  I've added it with 292435d → 2bb26ec.

Yes, this should be added.

Reply:
> 1. the object retrieved is not `application/vnd.oci.image.index.v1+json` type, it should be individual type created by discovery implementation.

But it *is* an `application/vnd.oci.image.index.v1+json`; with the `casEngines` proposal we'll update image-spec to add the new descriptor property.  And the point of the index template protocol is that consumers can recycle their `index.json` tooling, so we want the same type.

> 2. Example just paste the object content, needn't past curl command. The approaches are too widely.

Folks can certainly fetch this with tools other than cURL.  But I think it's a simple way to demonstrate fetching the index object from the expanded URI with content negotiation.